server/sonar-process/src/main/java/org/sonar/process/jmvoptions/JvmOptions.java

```diff
@@ -21,31 +21,90 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.process.MessageException;
 import org.sonar.process.Props;
 
+import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
 
 public class JvmOptions<T extends JvmOptions> {
   private static final String JVM_OPTION_NOT_NULL_ERROR_MESSAGE = "a JVM option can't be null";
 
+  private final HashMap<String, String> mandatoryOptions = new HashMap<>();
   private final LinkedHashSet<String> options = new LinkedHashSet<>();
 
-  public JvmOptions(String... mandatoryJvmOptions) {
-    Arrays.stream(requireNonNull(mandatoryJvmOptions, JVM_OPTION_NOT_NULL_ERROR_MESSAGE))
-      .forEach(this::add);
+  public JvmOptions() {
+    this(Collections.emptyMap());
+  }
+
+  public JvmOptions(Map<String, String> mandatoryJvmOptions) {
+    requireNonNull(mandatoryJvmOptions, JVM_OPTION_NOT_NULL_ERROR_MESSAGE)
+      .entrySet()
+      .stream()
+      .filter(e -> {
+        requireNonNull(e.getKey(), "JVM option prefix can't be null");
+        if (e.getKey().trim().isEmpty()) {
+          throw new IllegalArgumentException("JVM option prefix can't be empty");
+        }
+        requireNonNull(e.getValue(), "JVM option value can't be null");
+        return true;
+      }).forEach(e -> {
+        String key = e.getKey().trim();
+        String value = e.getValue().trim();
+        mandatoryOptions.put(key, value);
+        add(key + value);
+      });
   }
 
   public T addFromMandatoryProperty(Props props, String propertyName) {
```

Comment:
The term `mandatory` looks strange to me because by looking at the callers we use it with non mandatory property such as `.addFromMandatoryProperty(props, ProcessProperties.WEB_JAVA_ADDITIONAL_OPTS)`.
I would prefer to rename this method with addFromProperty.

Reply:
it's mandatory but not necessarily specified by the user thanks to a default value added at runtime.
This default value is not known to this class.